File: backend/pkg/auth/module.go
package auth

import (
	"errors"
	"time"

	apperrors "template/modules/core/pkg/errors"

	"github.com/gofiber/fiber/v2"
	"github.com/golang-jwt/jwt/v5"
	"gorm.io/gorm"
)

// Module implements the pluggable auth module interface
type Module struct {
	db            *gorm.DB
	jwtSecret     string
	jwtMiddleware fiber.Handler
	userProvider  *UserProvider
	roleProvider  *RoleProvider
}

// New creates a new Auth module instance
func New(db *gorm.DB, jwtSecret string) *Module {
	return &Module{
		db:            db,
		jwtSecret:     jwtSecret,
		jwtMiddleware: nil, // Will be set by SetJWTMiddleware
		userProvider:  NewUserProvider(db),
		roleProvider:  NewRoleProvider(db),
	}
}

// SetJWTMiddleware sets the JWT middleware for protected routes
func (m *Module) SetJWTMiddleware(middleware fiber.Handler) {
	m.jwtMiddleware = middleware
}

func (m *Module) UserProvider() *UserProvider {
	return m.userProvider
}

func (m *Module) RoleProvider() *RoleProvider {
	return m.roleProvider
}

// Name returns the module name
func (m *Module) Name() string {
	return "auth"
}

// RegisterRoutes registers authentication routes
func (m *Module) RegisterRoutes(router fiber.Router) {
	// Public auth routes
	auth := router.Group("/auth")
	auth.Post("/register", m.Register)
	auth.Post("/login", m.Login)
	auth.Post("/logout", m.Logout)

	// Protected auth routes
	if m.jwtMiddleware != nil {
		auth.Get("/me", m.jwtMiddleware, m.Me)
	} else {
		auth.Get("/me", m.Me) // Fallback without middleware
	}

	// CRUD routes for users
	users := router.Group("/users")
	users.Get("/", m.userProvider.ListHandler())
	users.Get("/schema", m.userProvider.SchemaHandler())
	users.Get("/:id", m.userProvider.GetHandler())
	users.Post("/", m.userProvider.CreateHandler())
	users.Put("/:id", m.userProvider.UpdateHandler())
	users.Delete("/:id", m.userProvider.DeleteHandler())

	roles := router.Group("/roles")
	roles.Get("/", m.roleProvider.ListHandler())
	roles.Get("/schema", m.roleProvider.SchemaHandler())
	roles.Get("/:id", m.roleProvider.GetHandler())
	roles.Post("/", m.roleProvider.CreateHandler())
	roles.Put("/:id", m.roleProvider.UpdateHandler())
	roles.Delete("/:id", m.roleProvider.DeleteHandler())
}

// Migrate runs database migrations
func (m *Module) Migrate(db *gorm.DB) error {
	if err := db.AutoMigrate(&User{}, &Role{}); err != nil {
		return err
	}

	for _, role := range defaultRoles() {
		if err := db.FirstOrCreate(&role, Role{Name: role.Name}).Error; err != nil {
			return err
		}
	}

	return nil
}

// RegisterRequest represents registration data
type RegisterRequest struct {
	Email     string `json:"email"`
	Password  string `json:"password"`
	FirstName string `json:"first_name"`
	LastName  string `json:"last_name"`
}

// LoginRequest represents login credentials
type LoginRequest struct {
	Email    string `json:"email"`
	Password string `json:"password"`
}

// AuthResponse represents auth response with token
type AuthResponse struct {
	Token string `json:"token"`
	User  *User  `json:"user"`
}

// Register creates a new user account
// POST /api/auth/register
func (m *Module) Register(c *fiber.Ctx) error {
	req := new(RegisterRequest)

	if err := c.BodyParser(req); err != nil {
		appErr := apperrors.BadRequest("Invalid request body")
		return c.Status(appErr.Code).JSON(appErr)
	}

	// Validation
	if req.Email == "" {
		appErr := apperrors.ValidationError(map[string]string{
			"email": "Email is required",
		})
		return c.Status(appErr.Code).JSON(appErr)
	}
	if req.Password == "" || len(req.Password) < 8 {
		appErr := apperrors.ValidationError(map[string]string{
			"password": "Password must be at least 8 characters",
		})
		return c.Status(appErr.Code).JSON(appErr)
	}

	// Check if user already exists
	var existingUser User
	if err := m.db.Where("email = ?", req.Email).First(&existingUser).Error; err == nil {
		appErr := apperrors.Conflict("User with this email already exists")
		return c.Status(appErr.Code).JSON(appErr)
	}

	// Create new user
	user := User{
		Email:     req.Email,
		Password:  req.Password, // Will be hashed by BeforeCreate hook
		FirstName: req.FirstName,
		LastName:  req.LastName,
		Role:      RoleUser,
		Active:    true,
	}

	if err := m.db.Create(&user).Error; err != nil {
		appErr := apperrors.Internal(err)
		return c.Status(appErr.Code).JSON(appErr)
	}

	// Generate JWT token
	token, err := m.generateToken(&user)
	if err != nil {
		appErr := apperrors.InternalWithMessage("Failed to generate token", err)
		return c.Status(appErr.Code).JSON(appErr)
	}

	// Set HTTP-only cookie
	m.setAuthCookie(c, token)

	return c.Status(201).JSON(AuthResponse{
		Token: token,
		User:  &user,
	})
}

// Login authenticates a user
// POST /api/auth/login
func (m *Module) Login(c *fiber.Ctx) error {
	req := new(LoginRequest)

	if err := c.BodyParser(req); err != nil {
		appErr := apperrors.BadRequest("Invalid request body")
		return c.Status(appErr.Code).JSON(appErr)
	}

	// Validation
	if req.Email == "" || req.Password == "" {
		appErr := apperrors.ValidationError(map[string]string{
			"email":    "Email is required",
			"password": "Password is required",
		})
		return c.Status(appErr.Code).JSON(appErr)
	}

	// Find user by email
	var user User
	if err := m.db.Where("email = ?", req.Email).First(&user).Error; err != nil {
		if errors.Is(err, gorm.ErrRecordNotFound) {
			appErr := apperrors.Unauthorized("Invalid email or password")
			return c.Status(appErr.Code).JSON(appErr)
		}
		appErr := apperrors.Internal(err)
		return c.Status(appErr.Code).JSON(appErr)
	}

	// Check if user is active
	if !user.Active {
		appErr := apperrors.Forbidden("Account is deactivated")
		return c.Status(appErr.Code).JSON(appErr)
	}

	// Verify password
	if !user.CheckPassword(req.Password) {
		appErr := apperrors.Unauthorized("Invalid email or password")
		return c.Status(appErr.Code).JSON(appErr)
	}

	// Generate JWT token
	token, err := m.generateToken(&user)
	if err != nil {
		appErr := apperrors.InternalWithMessage("Failed to generate token", err)
		return c.Status(appErr.Code).JSON(appErr)
	}

	// Set HTTP-only cookie
	m.setAuthCookie(c, token)

	return c.JSON(AuthResponse{
		Token: token,
		User:  &user,
	})
}

// Logout logs out the user
// POST /api/auth/logout
func (m *Module) Logout(c *fiber.Ctx) error {
	// Clear the auth cookie
	c.Cookie(&fiber.Cookie{
		Name:     "auth_token",
		Value:    "",
		Expires:  time.Now().Add(-time.Hour),
		HTTPOnly: true,
		Secure:   false, // Set to true in production with HTTPS
		SameSite: "Lax",
	})

	return c.Status(200).JSON(fiber.Map{
		"message": "Logged out successfully",
	})
}

// Me returns the current authenticated user
// GET /api/auth/me
func (m *Module) Me(c *fiber.Ctx) error {
	// Get user from context (set by JWT middleware)
	user := c.Locals("user").(*jwt.Token)
	claims := user.Claims.(jwt.MapClaims)
	userID := uint(claims["user_id"].(float64))

	// Fetch user from database
	var currentUser User
	if err := m.db.First(&currentUser, userID).Error; err != nil {
		if errors.Is(err, gorm.ErrRecordNotFound) {
			appErr := apperrors.NotFound("User")
			return c.Status(appErr.Code).JSON(appErr)
		}
		appErr := apperrors.Internal(err)
		return c.Status(appErr.Code).JSON(appErr)
	}

	return c.JSON(currentUser)
}

// generateToken creates a JWT token for the user
func (m *Module) generateToken(user *User) (string, error) {
	claims := jwt.MapClaims{
		"user_id": user.ID,
		"email":   user.Email,
		"role":    user.Role,
		"exp":     time.Now().Add(time.Hour * 24).Unix(), // 24 hours
		"iat":     time.Now().Unix(),
	}

	token := jwt.NewWithClaims(jwt.SigningMethodHS256, claims)
	return token.SignedString([]byte(m.jwtSecret))
}

// setAuthCookie sets the JWT token as an HTTP-only cookie
func (m *Module) setAuthCookie(c *fiber.Ctx, token string) {
	c.Cookie(&fiber.Cookie{
		Name:     "auth_token",
		Value:    token,
		Expires:  time.Now().Add(time.Hour * 24),
		HTTPOnly: true,
		Secure:   false, // Set to true in production with HTTPS
		SameSite: "Lax",
	})
}

File: backend-rust/src/module.rs
use std::sync::Arc;

use async_trait::async_trait;
use axum::{
    extract::State,
    http::{header, StatusCode},
    middleware::{self, Next},
    response::{IntoResponse, Response},
    routing::post,
    Json, Router,
};
use axum_extra::extract::CookieJar;
use axum_extra::extract::cookie::{Cookie, SameSite};
use chrono::Utc;
use jsonwebtoken::{decode, encode, DecodingKey, EncodingKey, Header, Validation};
use sea_orm::{
    ActiveModelTrait, ActiveValue::Set, ColumnTrait, DatabaseConnection, EntityTrait,
    QueryFilter,
};
use sea_orm_migration::MigratorTrait;
use serde::{Deserialize, Serialize};
use serde_json::json;

use backend_core::{crud_router, AppError, Module};

use crate::{
    entities::{
        role,
        user::{self, ActiveModel as UserActiveModel},
    },
    migration::Migrator,
    providers::{RoleProvider, UserProvider},
};

// ── JWT Claims ────────────────────────────────────────────────────────────────

#[derive(Debug, Serialize, Deserialize, Clone)]
pub struct Claims {
    pub user_id: i64,
    pub email: String,
    pub role: String,
    pub exp: i64,
    pub iat: i64,
}

// ── Shared state ──────────────────────────────────────────────────────────────

#[derive(Clone)]
struct AuthState {
    db: DatabaseConnection,
    jwt_secret: String,
}

// ── AuthModule ────────────────────────────────────────────────────────────────

pub struct AuthModule {
    db: DatabaseConnection,
    jwt_secret: String,
}

impl AuthModule {
    pub fn new(db: DatabaseConnection, jwt_secret: String) -> Self {
        Self { db, jwt_secret }
    }

    /// Axum middleware that validates JWT from Authorization header or cookie,
    /// and injects Claims into request extensions.
    pub fn require_auth() -> axum::middleware::FromFnLayer<
        impl Fn(axum::http::Request<axum::body::Body>, Next) -> std::pin::Pin<Box<dyn std::future::Future<Output = Response> + Send>>,
        (),
        (axum::http::Request<axum::body::Body>, Next),
    > {
        middleware::from_fn(auth_middleware)
    }
}

#[async_trait]
impl Module for AuthModule {
    fn name(&self) -> &str {
        "auth"
    }

    fn router(&self) -> Router {
        let state = Arc::new(AuthState {
            db: self.db.clone(),
            jwt_secret: self.jwt_secret.clone(),
        });

        let user_provider = Arc::new(UserProvider::new(self.db.clone()));
        let role_provider = Arc::new(RoleProvider::new(self.db.clone()));

        Router::new()
            // auth endpoints
            .route("/auth/register", post(register))
            .route("/auth/login", post(login))
            .route("/auth/logout", post(logout))
            .route("/auth/me", axum::routing::get(me)
                .route_layer(middleware::from_fn(auth_middleware)))
            // CRUD routes (usable without admin module)
            .nest("/users", crud_router(user_provider))
            .nest("/roles", crud_router(role_provider))
            .with_state(state)
    }

    async fn migrate(&self, db: &DatabaseConnection) -> Result<(), AppError> {
        Migrator::up(db, None).await.map_err(|e| AppError::Internal(e.into()))
    }
}

// ── Request / Response types ──────────────────────────────────────────────────

#[derive(Deserialize)]
struct RegisterRequest {
    email: String,
    password: String,
    first_name: Option<String>,
    last_name: Option<String>,
}

#[derive(Deserialize)]
struct LoginRequest {
    email: String,
    password: String,
}

#[derive(Serialize)]
struct AuthResponse {
    token: String,
    user: user::Model,
}

// ── Handlers ──────────────────────────────────────────────────────────────────

async fn register(
    State(state): State<Arc<AuthState>>,
    jar: CookieJar,
    Json(req): Json<RegisterRequest>,
) -> Result<impl IntoResponse, AppError> {
    if req.email.is_empty() {
        return Err(AppError::BadRequest("Email is required".into()));
    }
    if req.password.len() < 8 {
        return Err(AppError::BadRequest("Password must be at least 8 characters".into()));
    }

    // check duplicate
    if user::Entity::find()
        .filter(user::Column::Email.eq(&req.email))
        .one(&state.db)
        .await?
        .is_some()
    {
        return Err(AppError::Conflict("Email already in use".into()));
    }

    // default "user" role
    let default_role = role::Entity::find()
        .filter(role::Column::Name.eq("user"))
        .one(&state.db)
        .await?
        .ok_or_else(|| AppError::Internal(anyhow::anyhow!("Default role not found")))?;

    let hash = bcrypt::hash(&req.password, bcrypt::DEFAULT_COST)
        .map_err(|e| AppError::Internal(e.into()))?;

    let now = Utc::now();
    let new_user = UserActiveModel {
        email: Set(req.email),
        password: Set(hash),
        first_name: Set(req.first_name),
        last_name: Set(req.last_name),
        role_id: Set(default_role.id),
        active: Set(true),
        created_at: Set(now),
        updated_at: Set(now),
        ..Default::default()
    };
    let user = new_user.insert(&state.db).await?;

    let token = generate_token(&user, &default_role.name, &state.jwt_secret)?;
    let jar = set_auth_cookie(jar, &token);

    Ok((StatusCode::CREATED, jar, Json(AuthResponse { token, user })))
}

async fn login(
    State(state): State<Arc<AuthState>>,
    jar: CookieJar,
    Json(req): Json<LoginRequest>,
) -> Result<impl IntoResponse, AppError> {
    if req.email.is_empty() || req.password.is_empty() {
        return Err(AppError::BadRequest("Email and password are required".into()));
    }

    let user = user::Entity::find()
        .filter(user::Column::Email.eq(&req.email))
        .one(&state.db)
        .await?
        .ok_or_else(|| AppError::Unauthorized("Invalid email or password".into()))?;

    if !user.active {
        return Err(AppError::Forbidden("Account is deactivated".into()));
    }

    let matches = bcrypt::verify(&req.password, &user.password)
        .map_err(|e| AppError::Internal(e.into()))?;
    if !matches {
        return Err(AppError::Unauthorized("Invalid email or password".into()));
    }

    let role = role::Entity::find_by_id(user.role_id)
        .one(&state.db)
        .await?
        .ok_or_else(|| AppError::Internal(anyhow::anyhow!("Role not found")))?;

    let token = generate_token(&user, &role.name, &state.jwt_secret)?;
    let jar = set_auth_cookie(jar, &token);

    Ok((jar, Json(AuthResponse { token, user })))
}

async fn logout(jar: CookieJar) -> impl IntoResponse {
    let jar = jar.remove(Cookie::build("auth_token").path("/"));
    (jar, Json(json!({ "message": "Logged out successfully" })))
}

async fn me(
    State(state): State<Arc<AuthState>>,
    axum::Extension(claims): axum::Extension<Claims>,
) -> Result<Json<user::Model>, AppError> {
    let user = user::Entity::find_by_id(claims.user_id)
        .one(&state.db)
        .await?
        .ok_or_else(|| AppError::NotFound("User not found".into()))?;
    Ok(Json(user))
}

// ── Middleware ────────────────────────────────────────────────────────────────

async fn auth_middleware(
    jar: CookieJar,
    mut req: axum::http::Request<axum::body::Body>,
    next: Next,
) -> Response {
    // try Authorization header first, then cookie
    let token = req
        .headers()
        .get(header::AUTHORIZATION)
        .and_then(|v| v.to_str().ok())
        .and_then(|v| v.strip_prefix("Bearer "))
        .map(str::to_owned)
        .or_else(|| jar.get("auth_token").map(|c| c.value().to_owned()));

    let secret = req
        .extensions()
        .get::<Arc<AuthState>>()
        .map(|s| s.jwt_secret.clone());

    // If we can't get secret from extensions, try state
    let claims = token.zip(secret).and_then(|(tok, sec)| {
        decode::<Claims>(&tok, &DecodingKey::from_secret(sec.as_bytes()), &Validation::default())
            .ok()
            .map(|d| d.claims)
    });

    match claims {
        Some(c) => {
            req.extensions_mut().insert(c);
            next.run(req).await
        }
        None => (
            StatusCode::UNAUTHORIZED,
            Json(json!({ "error": "Authentication required" })),
        )
            .into_response(),
    }
}

// ── JWT helpers ───────────────────────────────────────────────────────────────

fn generate_token(user: &user::Model, role: &str, secret: &str) -> Result<String, AppError> {
    let claims = Claims {
        user_id: user.id,
        email: user.email.clone(),
        role: role.to_owned(),
        exp: (Utc::now() + chrono::Duration::hours(24)).timestamp(),
        iat: Utc::now().timestamp(),
    };
    encode(
        &Header::default(),
        &claims,
        &EncodingKey::from_secret(secret.as_bytes()),
    )
    .map_err(|e| AppError::Internal(e.into()))
}

fn set_auth_cookie(jar: CookieJar, token: &str) -> CookieJar {
    let cookie = Cookie::build(("auth_token", token.to_owned()))
        .path("/")
        .http_only(true)
        .same_site(SameSite::Lax)
        .max_age(time::Duration::hours(24))
        .build();
    jar.add(cookie)
}
